### ISSUE TITLE
Programming RPMB key, or how to avoid CFG_RPMB_WRITE_KEY=y

### ISSUE THREAD

**Issue by jenswi-linaro** [OP-TEE/optee_os#5338]
Hi all!

Problems with programing the RPMB key has been discussed recently in 
https://github.com/OP-TEE/optee_os/pull/5323 and https://github.com/OP-TEE/optee_docs/pull/157.

The config variable `CFG_RPMB_WRITE_KEY` must be set to `y` for OP-TEE to do this. But there are a few problems with this. From security point of view it may lead to a class break if ever such a signed OP-TEE binary is leaked. Then there's the accidents with writing unusable keys which in worst case may brick the device. Vendors will do well to never use `CFG_RPMB_WRITE_KEY=y`, but since the option is there it's a risk that it's just too convenient to use.

The best would be if this wasn't done by OP-TEE at all. If secure fuses etc used to derive the RPMB key are known the RPMB key can be calculated offline, possibly with help from a script for the last step. Typically the steps done by `huk_subkey_derive()` after the HUK has been retrieved with `tee_otp_get_hw_unique_key()`.

With the key calculated independent of OP-TEE we could let U-Boot and/or some user space application in Linux program the key during production. As a safety step to avoid accidents we can add a PTA function to verify that the calculated RPMB key matches OP-TEE's view. Providing such a function should not be a security problem since an eventual attacker just as well could do a brute force on the RPMB itself.

From testing point of view in CI we could add the capability to xtest to write the RPMB key if needed on QEMU. For real hardware I assume that the key will be programmed before it's added to a CI loop.

Thoughts, comments?

Thanks,
Jens

**Comment by jforissier** [comment]
> If secure fuses etc used to derive the RPMB key are known the RPMB key can be calculated offline

This may not always be true. If the device uses a derivation mechanism where the root key cannot leave the chip, then the key can only be generated by software running on the device (and probably only in secure world?). In this case some special firmware would be needed to program the RPMB key no matter what, be it a special build of OP-TEE or something else.

**Comment by jenswi-linaro** [comment]
OK. I had the impression that even if the root key doesn't leave the chip it would still be known because it was known how the chip was fused during production. But if some firmware randomizes and program this early in the life cycle of the platform then it will truly unknown to the outside.

In that case I guess we'd need a PTA function to extract the RPMB key too. This function should of course be protected with some other config flag and normally not enabled. Perhaps we can have some authentication scheme to make it obvious that it can't be used totally unprotected. I was hoping that we could avoid this.

**Comment by etienne-lms** [comment]
> > If secure fuses etc used to derive the RPMB key are known the RPMB key can be calculated offline
>
> This may not always be true.

Agree with @jforissier, leading to "...we'd need...":

> In that case I guess we'd need a PTA function to extract the RPMB key too.

That is almost what `CFG_RPMB_WRITE_KEY=y` does.
Do you think CFG_RPMB_WRITE_KEY is too sensitive to rely upon a config switch?

> For real hardware I assume that the key will be programmed before it's added to a CI loop.

I agree.
But Qemu is a bit specific in that, no?

**Comment by jenswi-linaro** [comment]
> > > If secure fuses etc used to derive the RPMB key are known the RPMB key can be calculated offline
> > 
> > 
> > This may not always be true.
> 
> Agree with @jforissier, leading to "...we'd need...":
> 
> > In that case I guess we'd need a PTA function to extract the RPMB key too.
> 
> That is almost what `CFG_RPMB_WRITE_KEY=y` does. Do you think CFG_RPMB_WRITE_KEY is too sensitive to rely upon a config switch?

`CFG_RPMB_WRITE_KEY=y` is very dangerous to use a on a production platform. With that you can extract the RPMB key without any authentication whatsoever. You'd need to be root or have control over tee-supplicant, but that's as far as security goes.

With a special function to extract the RPMB key it pretty clear that the function is very sensitive and can't be left without some level of authentication. `CFG_RPMB_WRITE_KEY=y` just works and there's a risk that it's overlooked when securing the system. As I said above, it can lead to a class breach of security.

> 
> > For real hardware I assume that the key will be programmed before it's added to a CI loop.
> 
> I agree. But Qemu is a bit specific in that, no?

Yes, but we could implement the secure world part in a secure way so serve as an example or at least to highlight a problem.

**Comment by ldts** [comment]
What I like about this proposal is that OP-TEE could return the key qualified: ie this is a development RPMB key vs this is a production RPMB key, and then let the user implement the policy. 

For instance, if the system is booting non-secure (ie, firmware signatures not being hw verified) the system might choose to still write the RPMB key. Under those circumstances it doesnt really matter if the key leaks, after all, the system hasnt been secured.

If the system is booting secure - firmware running signatures validated - asking for the key cant not lead to a leak.

**Comment by jenswi-linaro** [comment]
> @jenswi-linaro I was thinking about this a bit more.
> 
> Since the HUK is the only variable that is affecting the RPMB key calculation - all other inputs to the HUK are pretty much fixed for the platform - perhaps we should just declare the HUK as an _unique_id + property_ 
with the property being secure (ie, persistent) or non-secure.
> 
> And then force the user to be explicit about letting the non-secure HUK be used for the RPMB key generation - never allow it by default.


I'm sorry, I don't get it. We have the implementation of huk_subkey_derive() which basically MACs a usage enum plus some data using the HUK as key to the MAC. We depend on the HUK being secure. Where does non-secure come into this?

**Comment by koalo** [comment]
This thread and in particular the documentation https://optee.readthedocs.io/en/latest/architecture/secure_storage.html#security-considerations seem to imply that one should NOT use `CFG_RPMB_WRITE_KEY=y` to write the RPMB key. In particular,

> This configuration should only be enabled in a test or development environment.

So there must be a better way and the documentation also seems to mention a better way:

> The safest and most secure way is to program it from normal world during production in a secure environment.

But I do not understand how this could work, in particular on hardware using CAAM, where - in my understanding, which might be wrong - one can never know the HUK (to derive the PRMB key) without a secure world application calling the following code, not even during/before initial production or even by knowing the secure fuses as mentioned in the initial post.

https://github.com/OP-TEE/optee_os/blob/112396a58cf0d5d76456f5a959f5420fdcc241be/core/drivers/crypto/caam/blob/caam_blob.c#L59

So if we need anyway a secure world application to get the HUK to derive the RPMB, how could any solution be more secure than just using a secure world application with  `CFG_RPMB_WRITE_KEY=y` in the first place?

What do I overlook here?

**Comment by jenswikl** [comment]
> So if we need anyway a secure world application to get the HUK to derive the RPMB, how could any solution be more secure than just using a secure world application with `CFG_RPMB_WRITE_KEY=y` in the first place?
> 
> What do I overlook here?

`CFG_RPMB_WRITE_KEY=y` requires no authentication whatsoever and happily provides the RPMB key each time the normal world asks for it.

**Comment by koalo** [comment]
> CFG_RPMB_WRITE_KEY=y requires no authentication whatsoever and happily provides the RPMB key each time the normal world asks for it.

Yes, I am aware of that! But is there (can there be?) any reasonable alternative with CAAM hardware to "Having a separate OP-TEE binary in the factory" with `CFG_RPMB_WRITE_KEY=y` (which is according to https://optee.readthedocs.io/en/latest/architecture/secure_storage.html#security-considerations also a bad idea) that does not have a similar security issue?

**Comment by pblxptr** [comment]
I would like to bring another question in this topic as its quite related to programming RPMB key. Suppose that we hire external contractor to manufacture the boards, including board provisioning and so on. Let's say that platform features some security mechanisms that allow us to program some sensitive keys, fuses and so on securely, even in untrusted remote manufacturer's environment. The only security hole that is left is programming RPMB key. I'm wondering is there any way to program this RPMB key securely in contractor's env when -  given that upon programming procedure - the key is being sent to eMMC in "plaintext"?

**Comment by jenswikl** [comment]
As I said above:

> In that case I guess we'd need a PTA function to extract the RPMB key too. This function should of course be protected with some other config flag and normally not enabled. Perhaps we can have some authentication scheme to make it obvious that it can't be used totally unprotected. I was hoping that we could avoid this.

This depends on trusting the provisioning process.

**Comment by koalo** [comment]
> In that case I guess we'd need a PTA function to extract the RPMB key too. This function should of course be protected with some other config flag and normally not enabled.

The question is: Why is that more secure than `CFG_RPMB_WRITE_KEY=y` ?

**Comment by jenswikl** [comment]
Because the PTA will only provide the key after the normal world has authenticated itself. The authentication could be a challenge response via a secure server, only available in a secure environment.

**Comment by koalo** [comment]
> Because the PTA will only provide the key after the normal world has authenticated itself. The authentication could be a challenge response via a secure server, only available in a secure environment.

Ok, yes! With challenge response with asymmetric encryption and the private key stored in a separate instance (i.e. not in a program running on the same device), this should work.

Is there already a PTA function with a similar authentication mechanism implemented somewhere in OP-TEE to get started?

**Comment by jenswikl** [comment]
I'm afraid not. The closest thing we have is the call to `shdr_verify_signature2()` in `ree_fs_ta_open()`. It can be a good idea to consider fault mitigation for this call. If fault mitigation becomes too big a step for this feature, we can do that later.
